layer/event_logging.h

```diff
@@ -109,9 +108,11 @@ class Event {
 
   LogLevel GetLogLevel() const { return log_level_; }
 
+ protected:
+  std::vector<Attribute *> attributes_;
```

Comment:
Instead of making the variable protected, what do you think about adding a protected function `initAttributes`? This should be more greppable and easier to debug because all updates will be in the same function